Improve embedding_bag add kernel

jamesr66a opened the pull request:
This PR makes it so that we call into the high-performance `EmbeddingLookup` function from C2 within embedding_bag for `add` mode. This is highly-optimized and does dynamic dispatch based on cpuid. It also makes it so that we elide creating the offset2bag tensor in the fast path case. That operation was contributing a significant portion of runtime to the operator.

Benchmark script (with a hack to make jit not DCE/CSE): https://gist.github.com/jamesr66a/73baed47400dcf2221bad996c4b57782

=== Baseline (master) === 

```
time_per_iter 8.726580142974853e-05
GB/s 4.588097438402839
```

== Test (this PR) ===

```
time_per_iter 2.5180697441101074e-05
GB/s 15.900433295643158
```

dzhulgakov commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -52,13 +57,92 @@ static void index_select_add(const Tensor &select_indices,
-            src_data + src_stride0 * select_indices_data[i], src_stride1,
-            output_data + output_stride0 * add_indices_data[i], output_stride1);
+  auto accessor = offsets.accessor<int64_t, 1>();
+  std::vector<int> lengths(offsets.numel());
```
wait, you don't use lengths here at all, do you?

jamesr66a replied:
This is a WIP. I'm trying to get the FBCode build disaster fixed before i start fixing the kernel implementation

dzhulgakov commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -52,13 +57,92 @@ static void index_select_add(const Tensor &select_indices,
+      lengths.push_back(accessor[i] - lower);
+      lower = accessor[i];
+    }
+    lengths.push_back(numel - lower);
```
maybe add a TODO to just changed the underlying kernel to take offsets, not lengths?

dzhulgakov commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -84,13 +168,25 @@ static void index_select_scale_add(const Tensor &select_indices,
+  // We specialize the case where both the src and output are contiguous in the
+  // last dimension to give the compiler a better chance to vectorize.
+  if (src_stride1 == 1 && output_stride1 == 1) {
+    for (int64_t i = 0; i < numel; i++) {
```
EmbeddingLookup supports weighted version too - see `weights` argument, so let's call it here

dzhulgakov commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -59,6 +64,59 @@ static void index_select_add(const Tensor &select_indices,
+    auto accessor = offsets.accessor<int64_t, 1>();
+    std::vector<int> lengths;
+
+    int64_t lower = accessor[0];
```
it assumes the offsets is non-empty. Probably usually the case, but better to work on empty batch too (seems like we'd need to fix functional.py too)

dzhulgakov commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -248,9 +366,9 @@ _embedding_bag_cpu(const Tensor &weight, const Tensor &indices,
         AT_ASSERT(mode == MODE_SUM);
         index_select_scale_add<scalar_t>(
-            indices, offset2bag, per_sample_weights, weight, output);
+            indices, offset2bag, per_sample_weights, weight, output, offsets);
```
you can probably save even more compute by skipping make_offset2bag vector in case we specialize here. Otherwise derivatives of offsets twice.

at least - add a TODO here so that the next person knows it

jamesr66a replied:
Uh, I think i'm gonna do that. make_offset2bag takes a ton of runtime in my tests and if we can get rid of it we can probably squeeze more perf out of this. Thanks for the tip

Commit:
Improve embedding_bag add kernel (#19329)

Summary:
This was actually getting pretty poor throughput with respect to memory bandwidth. I used this test to measure the memory bandwidth specifically for the AXPY call: https://gist.github.com/jamesr66a/b27ff9ecbe036eed5ec310c0a3cc53c5

And I got ~8 GB/s before this change, but ~14 GB/s after this change.

This seems to speed up the operator overall by around 1.3x (benchmark: https://gist.github.com/jamesr66a/c533817c334d0be432720ef5e54a4166):

== Before ==

time_per_iter 0.0001298875093460083
GB/s 3.082544287868467

== After ==

time_per_iter 0.00010104801654815674
GB/s 3.9623142905451076

The large difference between the local BW increase and the full-op BW increase likely indicates significant time is being spent elsewhere in the op, so I will investigate that.

EDIT: I updated this PR to include a call into caffe2/perfkernels. This is the progression:

before

time_per_iter 8.983819484710693e-05
GB/s 4.456723564864611

After no axpy
time_per_iter 7.19951868057251e-05
GB/s 5.56126065872172

AFter perfkernels
time_per_iter 5.6699180603027346e-05
GB/s 7.061548257694262

After perfkernels no grad
time_per_iter 4.388842582702637e-05
GB/s 9.122769670026413
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19329

Differential Revision: D14969630

fbshipit-source-id: 359b48eac218463d4ff13bdf22d31c70bf35281d

zou3519 commented on test/test_nn.py:
```diff
@@ -2563,12 +2563,15 @@ def test_contig_wrong_stride_cudnn(self):
-        self._test_EmbeddingBag(False, 'sum', True)
-        self._test_EmbeddingBag(False, 'mean', True)
+        for dtype in [torch.double, torch.float]:
+            # TODO: figure out why backward on float breaks
```
I don't think it's "broken" on backward; it's just that the precision on the test needs to be adjusted because the gradients are so large. If this hypothesis is true, then test_embedding_bag should fail for `mode=sum` but not fail for `mode={max,mean}`.

jamesr66a replied:
@zou3519 IME it fails for all off sum, max, and mean

zou3519 replied:
What is the magnitude of the gradient for "max" and "mean"? If it is on the order of magnitude of 1 then there are definitely precision issues, then

cpuhrsch commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -95,6 +158,67 @@ static void index_select_scale_add(const Tensor &select_indices,
+                                          const Tensor& offsets) {
+  int64_t ddim = src.size(1);
+  auto* scale_data = scale.data<float>();
+  auto select_indices_data = select_indices.data<int64_t>();
```
Do we actually assert that these data types match somewhere?

jamesr66a replied:
@cpuhrsch i think this does it: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/EmbeddingBag.cpp#L214

Also there's a PR out to fix that location: https://github.com/pytorch/pytorch/pull/19432

cpuhrsch commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -5,6 +5,8 @@
 
 #include <TH/THBlasUtils.h>
 
+#include <caffe2/perfkernels/embedding_lookup.h>
```
How does the dispatch work? Keep in mind that this file isn't compile with avx.

dzhulgakov replied:
It's a similar mechanism to aten/DISPATCH and does the right thing with different compiler options and runtime cpuid dispatch between them: https://github.com/pytorch/pytorch/blob/master/caffe2/perfkernels/common.h

we should bring them together, but it's for another day

soumith commented:
this also calls caffe2::EmbeddingLookup for the GPU. Is the perf of the caffe2 kernel better or equal on GPU? (I remember we optimized the GPU kernel quite a bit on the ATen side)

cpuhrsch commented:
@jamesr66a - is it worthwhile to try a few more input shapes and number of threads for the benchmark?

jamesr66a commented:
@soumith does this actually get called on GPU? I thought the only callsites for `index_select_add` and `index_select_scale_add` were in `_embedding_bag_cpu`

soumith commented:
@jamesr66a great, it was my bad not to check that.
